Refactored the document, implemented todos (#16)
rarimo/docs: docs/products/polygon-id-state-replication/tutorial-verifying-polygon-id-credentials-in-ethereum-dapp.mdx

---
title: Polygon ID State Replication Usage
---

# Tutorial: Verifying Polygon ID Credentials in Ethereum DApp

In this tutorial, we will build the DApp that mints an NFT if the user was born before some date. Given an issuer anchored to Polygon, we can replicate its state to another chain(in our case, Ethereum) to use zero-knowledge proofs in a DApp.

## User flow

1. The front-end creates the QR code with proof request from the user and shows it on the page.

  ![Proof request QR-code](/img/polygon-id-integration-qr-code.png)

1. The user can scan this QR code with his Polygon ID Wallet and generate proof that he possesses valid credentials.

  ![Proof generated, connect wallet to proceed](/img/polygon-id-integration-generated.png)

1. After that, the relayer will be called to replicate the state from Polygon to Ethereum.

  ![Performing state transition](/img/polygon-id-integration-relay.png)

> 1. After the successful verification, we mint an SBT for the user.

Reviewer: ```suggestion
1. User can submit the proof with his MetaMask and mint the SBT
```

Keep the user flow explanations user-centric

        const zkProof = zkProofPayload.body.scope[0] as ZKProof

        const txBody = getProveIdentityTxBody(
            '1',
            zkProof.pub_signals.map(el => BigInt(el)),
            [zkProof.proof.pi_a[0], zkProof.proof.pi_a[1]],
            [
            [zkProof.proof.pi_b[0][1], zkProof.proof.pi_b[0][0]],
            [zkProof.proof.pi_b[1][1], zkProof.proof.pi_b[1][0]],
            ],
            [zkProof.proof.pi_c[0], zkProof.proof.pi_c[1]],
        )

        const tx = await provider?.signAndSendTx?.({
            to: config?.[
                `QUERY_VERIFIER_CONTRACT_ADDRESS_${selectedChainToPublish}`
            ],
            ...txBody,
        })

        verificationSuccessTx.set((tx as EthTransactionResponse).transactionHash)

        navigate(RoutesPaths.authSuccess)
        } catch (error) {
        ErrorHandler.process(error)
        }

        setIsSubmitting(false)
    }, [
        getProveIdentityTxBody,
        jwzToken,
        navigate,
        provider,
        selectedChainToPublish,
        verificationSuccessTx,
    ])

/* ... */
```

Now you're ready to use the Polygon ID State Replication with Rarimo!

> If you want to proceed further, we recommend you to clone the front-end from the [GitHub](https://github.com/rarimo/web-client-polygonid/tree/main), make the changes you need and use it for your purposes.

Reviewer: Instead of "proceed further" - "tweak or expand this example", also add a link to contracts
